contrib/babelfishpg_tsql/sql/sys_functions.sql

```diff
@@ -3788,3 +3788,246 @@ RETURNS sys.SYSNAME AS $$
 $$
 LANGUAGE plpgsql IMMUTABLE
 STRICT;
+
+
+-- Another definition of date_bucket() with arg PG_CATALOG.TEXT since ANYELEMENT cannot handle type unknown.
+CREATE OR REPLACE FUNCTION sys.date_bucket(IN datepart PG_CATALOG.TEXT, IN number INTEGER, IN date PG_CATALOG.TEXT, IN origin PG_CATALOG.TEXT default NULL) RETURNS PG_CATALOG.TEXT 
+AS 
+$body$
+DECLARE
+BEGIN
+    IF datepart NOT IN ('year', 'quarter', 'month', 'week', 'day', 'hour', 'minute', 'second', 'millisecond') THEN
+        RAISE EXCEPTION '% is not a recognized date_bucket option.', datepart;
+    ELSIF number IS NULL THEN
+        RAISE EXCEPTION 'Argument data type NULL is invalid for argument 2 of date_bucket function.';
+    ELSIF date IS NULL THEN
+        RAISE EXCEPTION 'Argument data type NULL is invalid for argument 3 of date_bucket function.';
+    ELSE
+        RAISE EXCEPTION 'Argument data type % is invalid for argument 3 of date_bucket function.', pg_typeof(date);
+    END IF;
+END;
+$body$ 
+LANGUAGE plpgsql IMMUTABLE;
+
+CREATE OR REPLACE FUNCTION sys.date_bucket(IN datepart PG_CATALOG.TEXT, IN number INTEGER, IN date ANYELEMENT, IN origin ANYELEMENT default NULL) RETURNS ANYELEMENT 
+AS 
```

Comment:
Can there be a case where date is anyelement and origin is text or date is text and origin is anyelement . If yes then how are we handling it ?

Reply:
Created a new definition of function date_bucket() where date is of type anyelement and origin is of type text. By that we can throw exception with appropriate message about argument 3 and argument 4. Added test cases for the same.